#162606962 Serve landing page

darthrighteous opened the pull request:
#### What does this PR do?
Serve the landing page

#### Description of Task to be completed?
- set up actions and reducer to handle fetching article by category
- ensure landing page loads data from the backend

#### How should this be manually tested?
- check out to this branch `feature/162606962/serve-landing-page`
- run `npm start` to start up the server
- open `localhost:3000` to view the landing page
- use redux dev tool to observe the state and state history

#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
[162606962](https://www.pivotaltracker.com/story/show/162606962)

#### Screenshots (if appropriate)

#### Questions:
N/A

#### Link to Review app:
[review app](https://valinor-ah-frontend-stag-pr-26.herokuapp.com/)

Andela-Jalil committed:
feat(homePage): serve landing page

- connect landing page to display data from the backend

[Finishes #162606962]

KvNGCzA commented on src/reducers/articlesReducer.js:
```diff
@@ -0,0 +1,22 @@
+        [categoryName]: payload
+      });
+    default:
+      return state;
```
If you set the initial state to {}. Wouldn't all the default values in the store be cleared out?

darthrighteous replied:
No, this "state" is only scoped to the `articleReducer`

KvNGCzA replied:
Noted

KvNGCzA commented on src/reducers/articlesReducer.js:
```diff
@@ -0,0 +1,22 @@
+  switch (type) {
+    case FETCH_CATEGORY_REQUEST:
+      return ({
+        ...state,
```
same as line 18 below

darthrighteous replied:
see reply below

KvNGCzA commented on src/reducers/articlesReducer.js:
```diff
@@ -0,0 +1,22 @@
+    case FETCH_CATEGORY_FAILURE:
+      return ({
+        ...state,
+        [categoryName]: payload
```
same as line 18 below

darthrighteous replied:
see reply below

KvNGCzA commented on public/styles/HomePage.scss:
```diff
@@ -20,13 +20,28 @@
+        float: left;
+        padding: 0 0 0 2em;
+        text-align: center;
+        width: 78%;
```
The properties from line 23 - 26, except text-align have already been set in index.scss lines 45 - 47. So you can remove these.

darthrighteous replied:
You're right. I've fixed this now

KvNGCzA commented on public/styles/HomePage.scss:
```diff
@@ -20,13 +20,28 @@
+            position: relative;
+            text-align: left;
+            top: 0;
+            transition: all 0.5s;
```
The properties from line 40 to 44 except text-align have been set in index.sccs  lines 70 - 73. So you can remove these.

darthrighteous replied:
done

KvNGCzA commented on src/components/home/HomePage.js:
```diff
@@ -1,15 +1,68 @@
+  });
+};
+
+CardListArray.propTypes = {
```
This works, but I think it should be moved to the bottom of the file, next to HomePage.propTypes

darthrighteous replied:
Done.

KvNGCzA commented on src/components/home/HomePage.js:
```diff
@@ -1,15 +1,68 @@
+  articlesByCategory: PropTypes.object.isRequired
+};
+
+export class HomePageComponent extends Component {
```
Why was this component renamed?

darthrighteous replied:
This was renamed to have a different name from the default export which is `HomePage` as well

KvNGCzA replied:
Seen

sulenchy commented:
The review app is not up and running. Can you do something about it?

darthrighteous commented:
@sulenchy Check it again now

sulenchy commented on src/components/Card.js:
```diff
@@ -12,8 +12,8 @@ export const Card = (props) => {
-    <div className="card" style={{ backgroundImage: `url(${backgroundImage})` }}>
-      <a href={link}>
+    <div className="card" style={{ backgroundImage: 'url(https://bit.ly/2B4NZ51)' }}>
+      <NavLink to={`/articles/${link}`}>
```
Should we be using the slug to fetch a single article or id? If yes, hope it won't amount to extra work going forward?

darthrighteous replied:
This line here does not infer what we're using to fetch articles from the backend. However, I believe we are using id not slug.

sulenchy commented on src/components/home/HomePage.js:
```diff
@@ -1,15 +1,64 @@
+  constructor(props) {
+    super(props);
+    this.state = {
+      articles: mockArticles
```
mockArticles? are we fetching from database or from mockData?

darthrighteous replied:
This is used to populate the popular posts sidebar. There is currently no way to fetch the most popular article from the database. or no way to even know which articles are the most popular. So I left this as mock data.

The only alternative I see here is to replace this sidebar with most recent posts. I was planning to bring this up in our next sync/TTL interaction

Andela-Jalil committed:
feat(homePage): implement popular posts fetch

- add spinner and global state, reducer, actions
- add action to fetch popular posts from the backend and serve landing page

[#162606962]

mbilesanmi commented on webpack.common.js:
```diff
@@ -75,6 +75,7 @@ module.exports = {
   },
   plugins: [
     new webpack.DefinePlugin({
+      'process.env.NODE_ENV': JSON.stringify(process.env.NODE_ENV),
```
Are we using this anywhere?

darthrighteous replied:
No, I will remove this.

darthrighteous replied:
done

mbilesanmi commented on src/store/initialState.js:
```diff
@@ -3,5 +3,7 @@ export default {
-  }
+  },
+  articlesByCategory: {},
+  popular: []
```
popular what is this?

darthrighteous replied:
I'll rename this to `popularArticles`

darthrighteous replied:
done

mbilesanmi commented on src/reducers/popularPostsReducer.js:
```diff
@@ -0,0 +1,20 @@
+  switch (type) {
+    case FETCH_POPULAR_SUCCESS:
+    case FETCH_POPULAR_FAILURE:
+      return ([
```
Is this always the case for both success and failure? Would we have `...payload.articles` in all instances?

darthrighteous replied:
No this implementation is wrong. I will fix this

darthrighteous replied:
fixed

mbilesanmi commented on src/components/home/PopularPosts.js:
```diff
@@ -5,29 +5,38 @@ import { formatDate } from '../../utils';
-              </span>
-            </div>
-          )) }
+          { articles.slice(0, 6).map((currentArticle, index) => {
```
Why are we slicing here?

darthrighteous replied:
It only shows the top 6 most popular articles

mbilesanmi replied:
How many are we fetching from the backend? Did we set a limit to only return 6 or all?

mbilesanmi replied:
@Darthrighteous ☝️

darthrighteous replied:
@mbilesanmi it only returns 6 from the backend

darthrighteous replied:
@mbilesanmi which means there is no reason to still be slicing, I have removed it now.

mbilesanmi commented on src/components/home/HomePage.js:
```diff
@@ -1,20 +1,55 @@
+  const categories = Object.entries(articlesByCategory);
+
+  return categories.map((category) => {
+    if (category[1].articles) {
```
Why do we have the logic like it is? Is category[0] always gonna be the title? What is category[1]? Why return an error message as you have it below? Can you upload a screenshot of how it would look if one category has an error while others are okay?

Also, upload a screenshot of the categories response.

darthrighteous replied:
category[0] will always be the title, category[1] is an object containing the articles in that category.

This is how the errors are displaying
<img width="777" alt="screen shot 2018-12-18 at 4 14 06 pm" src="https://user-images.githubusercontent.com/31810086/50163126-1339dd80-02e0-11e9-8ab8-b18f247b5c11.png">

darthrighteous replied:
I have now changed it to use the toaster instead
<img width="938" alt="screen shot 2018-12-18 at 4 22 42 pm" src="https://user-images.githubusercontent.com/31810086/50164667-9d377580-02e3-11e9-9ded-8a6e32af536f.png">

mbilesanmi reviewed (changes requested):
Hi @Darthrighteous. Please review my comments ASAP.

Thanks.
